Alerting: Enable Alert rule tags to override PagerDuty Severity setting

AndrewBurian opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/master/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the master branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**: 

Uses alert tags to optionally override the `severity` in the PagerDuty event.

This gives you the ability to send events of different severity to the same service without needing a separate notification channel configured.

AndrewBurian committed:
Use alert tags to override PD Severity setting

AndrewBurian committed:
Minor doc clarification

oddlittlebird commented on docs/sources/alerting/notifications.md:
```diff
@@ -105,7 +105,7 @@ Integration Key | Integration key for PagerDuty.
 Auto resolve incidents | Resolve incidents in PagerDuty once the alert goes back to ok
 
-**Note:** The tags `Class`, `Group`, and `Component` have special meaning in the [Pagerduty Common Event Format - PD-CEF](https://support.pagerduty.com/docs/pd-cef).  If an alert panel defines these tag keys they will be transposed to the root of the event sent to Pagerduty.  This means they will be available within the Pagerduty UI and Filtering tools.
+**Note:** The tags `Severity`, `Class`, `Group`, and `Component` have special meaning in the [Pagerduty Common Event Format - PD-CEF](https://support.pagerduty.com/docs/pd-cef).  If an alert panel defines these tag keys they will be transposed to the root of the event sent to Pagerduty.  This means they will be available within the Pagerduty UI and Filtering tools. A Severity tag set on an alert  ...
```
```suggestion
**Note:** The tags `Severity`, `Class`, `Group`, and `Component` have special meaning in the [Pagerduty Common Event Format - PD-CEF](https://support.pagerduty.com/docs/pd-cef). If an alert panel defines these tag keys, then they are transposed to the root of the event sent to Pagerduty. This means they will be available within the Pagerduty UI and Filtering tools. A Severity tag set on an alert overrides the global Severity set on the notification channel if it's a valid level.
```

AndrewBurian committed:
Doc wording

Co-Authored-By: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

AndrewBurian commented:
I somehow managed to remove @grafana/backend-platform from the reviewers list... My bad. Can someone re-add that?

bergquist reviewed (changes requested):
Hi

How is this better than using different notification channels? From my perspective that would achive the same thing.

AndrewBurian commented:
It's both cleaner, and enables better division of responsibility.

In our use case, the notification channels are configured by the SRE team, and then consumed at the dashboard level by the engineers. We currently run 38 services, times 4 severity levels, thats crossing 150 notification channels that need to be manually added, and then searched for in the alerting panel. If a service changes or needs a new integration key, that's 4 copies of the same service that need updated.

I really like the addition of alert tags, and using that control to add things to the PagerDuty event fields for Class, Group, and Component has already been useful. Even if the configuration complexity wasn't being 4x'd without it, I would still argue that it's a good use of tags to bring that control to a more granular level per-alert rather than per-channel.

bergquist commented on pkg/services/alerting/notifiers/pagerduty_test.go:
```diff
@@ -197,6 +197,7 @@ func TestPagerdutyNotifier(t *testing.T) {
 						{Key: "group", Value: "aGroup"},
 						{Key: "class", Value: "aClass"},
 						{Key: "component", Value: "aComponent"},
+						{Key: "severity", Value: "warning"},
```
Could you please add a test using another severity as well?

bergquist reviewed (changes requested):
I agree that this is cleaner. This is something we want to achieve at a generic level rather then just for one alert notifier. But that's another issue :)

Could you please add more tests with different severity levels so that we test that it's based on input data. Once that is done I'll merge this.

AndrewBurian committed:
Additional testing

One additional case for custom severity info.
Another case for an invalid PD severity, will keep the custom data
but default to the alert channel default severity.

bergquist committed:
Merge branch 'master' into pd-severity-tag

* master: (98 commits)
  NewPanelEdit: Refactor value mappings UI to work better with new panel edit (#22808)
  FieldOverrides: Move FieldConfigSource from fieldOptions to PanelModel.fieldConfig (#22600)
  Reporting: Update docs with correct logger name (#22892)
  Design tweaks (#22886)
  Remove duplicated localStorage mock (#22872)
  Rich history UX fixes (#22783)
  Storybook: Solve deployment issues (#22873)
  Docs: Update templating.md (#22881)
  AzureMonitor: support workspaces function for template variables (#22882)
  Check if the datasource is of type loki using meta.id instead of name. (#22877)
  Prometheus: Render missing labels in legend formats as an empty string (#22355)
  BarGauge: Fix strict null that breaks Storybook build (#22871)
  SQLStore: Add migration for adding index on annotation.alert_id (#22876)
  Docs: Update export-pdf.md (#22767)
  Variables: adds missing feature toggle in DashboardModel (#22868)
  Devenv: adds grafana block with a customizeable version (#22867)
  Alerting: support alerting on data.Frame (that can be time series) (#22812)
  AlertNotifications: Translate notifications IDs to UIDs in Rule builder (#19882)
  CircleCI: Don't build Storybook on PR (#22865)
  Graphite: Rollup Indicator (#22738)
  ...

bergquist commented:
Thanks! I'll merge this one the build passes :)

bergquist commented:
Depends on https://github.com/grafana/grafana/pull/22909

bergquist committed:
Merge branch 'master' into pd-severity-tag

* master: (27 commits)
  Upgrade golangci-lint and fixes some linting errors. (#22909)
  Storybook: Remove unneccesary stories and reorganize a bit (#22959)
  Menu: aligning user menu options  (#22940)
  Select: Fixed select text positition (#22952)
  CircleCI: Pin grabpl to v0.2.0 (#22906)
  Docs: fix broken link (#22949)
  Variables: migrates ad hoc variable type to react/redux.  (#22784)
  AuthProxy: Fixes bug where long username could not be cached (#22926)
  Admin: Added Name to user list (#22939)
  Docs: Update email notifier documentation (#22919)
  update latest to 6.7.1 (#22933)
  changelog 6.7.1 (#22932)
  NewPanelEdit: Organise sidebar in tabs  (#22870)
  Query components unsafe lifecycle methods (#21163)
  Fix typo (#22923)
  Panels: Fixed size issue with panels when existing panel edit mode (#22912)
  Azure: Fixed dropdowns not showing current value (#22914)
  Reporting: Update docs with new default value (#22916)
  Fix: failing tests when running on none english os. (#22913)
  BackendSrv: only add content-type on POST, PUT requests (#22910)
  ...

bergquist commented:
Thank you for contributing!

AndrewBurian commented:
@oddlittlebird FYI, I think the documentation changes here are in the grafana docs under v6.7, but the code is not actually present in the v6.7.3 release, so those are out of sync. Judging by the milestone here, I think the docs aren't meant to show up until v7

oddlittlebird commented:
@bergquist, could you check on this?
